src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -120,11 +120,51 @@ void CoreWorkerDirectTaskSubmitter::OnWorkerIdle(const WorkerAddress &addr,
 }
 
 void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
-    const TaskSpecification &resource_spec) {
+    const TaskSpecification &resource_spec, const rpc::Address *address) {
+  RAY_CHECK(resource_spec.GetMessage().task_id().size() > 0);
   if (worker_request_pending_) {
     return;
   }
-  RAY_CHECK_OK(lease_client_.RequestWorkerLease(resource_spec));
+
+  WorkerLeaseInterface &lease_client = lease_client_;
+  if (address && address->raylet_id() != "") {
+    // Connect to raylet.
+    ClientID raylet_id = ClientID::FromBinary(address->raylet_id());
+    auto it = remote_lease_clients_.find(raylet_id);
+    if (it == remote_lease_clients_.end()) {
+      RAY_LOG(DEBUG) << "Connecting to raylet " << raylet_id;
+      it = remote_lease_clients_.emplace(raylet_id, lease_client_factory_(*address)).first;
+    }
+    RAY_LOG(DEBUG) << "Sending " << resource_spec.TaskId() << " to raylet " << raylet_id;
+    lease_client = *it->second;
+  }
+
+  RAY_CHECK_OK(lease_client.RequestWorkerLease(
+      resource_spec,
+      [this, resource_spec](const Status &status, const rpc::WorkerLeaseReply &reply) {
+        RAY_CHECK(resource_spec.GetMessage().task_id().size() > 0);
```

Comment:
@ericl, this check is failing right now. Do you have any ideas about this?

Reply:
The message is mutable, so it gets swapped out earlier. You have to capture any fields you care before RequestNewWorker returns.

Reply:
By the way, you should probably rebase on https://github.com/ray-project/ray/pull/6147 which cleans up a bunch of this code.

Reply:
Hmm I'm not sure I get it...shouldn't capturing `resource_spec` copy it and then keep the shared_ptr underneath in scope? What should I be capturing to make this work?

Thanks, I'll try rebasing.

Reply:
Nope. The `const` is misleading. You should think of task spec as `shared_ptr<rpc::TaskSpec>`, which can be mutated (or swapped out). So if you look for "Swap" in this file, you'll see that the TaskSpecification is moved shortly after this function is called.

I guess this is kind of unfortunate since this means you can't use resource_spec again. You could fix this by deferring the Swap here: https://github.com/ray-project/ray/blob/6c66bf1df6a0f0f21e0b869c48b5ed75103aa11b/src/ray/core_worker/transport/direct_task_transport.cc#L193

Until the task is actually ready to be sent.

Reply:
I see, this is pretty unfortunate. I guess we'll have to worry about this later in a larger refactoring, but we really should not allow objects to get mutated like this. It makes it quite difficult to use lambdas correctly.

I think for now, I'll just copy the task data, since it seems like it'll be pretty complicated to figure out when it's safe to swap.